tests/integration/src/test/java/org/mule/context/notification/FlowStackTestCase.java

```diff
@@ -0,0 +1,603 @@
+/*
+ * Copyright (c) MuleSoft, Inc.  All rights reserved.  http://www.mulesoft.com
+ * The software in this package is published under the terms of the CPAL v1.0
+ * license, a copy of which has been included with this distribution in the
+ * LICENSE.txt file.
+ */
+package org.mule.context.notification;
+
+import static org.hamcrest.CoreMatchers.is;
+import static org.hamcrest.CoreMatchers.not;
+import static org.hamcrest.CoreMatchers.nullValue;
+import static org.hamcrest.CoreMatchers.startsWith;
+import static org.hamcrest.collection.IsCollectionWithSize.hasSize;
+import static org.hamcrest.core.IsCollectionContaining.hasItem;
+import static org.junit.Assert.assertThat;
+
+import org.mule.DefaultMuleMessage;
+import org.mule.api.MuleEvent;
+import org.mule.api.MuleException;
+import org.mule.api.client.LocalMuleClient;
+import org.mule.api.config.MuleProperties;
+import org.mule.api.context.notification.FlowCallStack;
+import org.mule.api.context.notification.FlowStackElement;
+import org.mule.api.context.notification.MessageProcessorNotificationListener;
+import org.mule.api.processor.MessageProcessor;
+import org.mule.tck.junit4.FunctionalTestCase;
+import org.mule.tck.junit4.rule.SystemProperty;
+
+import java.util.List;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.TimeUnit;
+
+import org.junit.Before;
+import org.junit.Rule;
+import org.junit.Test;
+
+public class FlowStackTestCase extends FunctionalTestCase
+{
+
+    public static class FlowStackAsserter implements MessageProcessor
+    {
+
+        public static FlowCallStack stackToAssert;
+        public static CountDownLatch stackLatch;
+
+        @Override
+        public MuleEvent process(MuleEvent event) throws MuleException
+        {
+            stackToAssert = new DefaultFlowCallStack(event.getFlowCallStack());
+            if (stackLatch != null)
```

Comment:
Is not clear to me why in some case you need synchronization and no in others. Can you make that more explicit? (maybe using two different components)